features/mbedtls/targets/TARGET_NUVOTON/TARGET_NUC472/des/des_alt.c

```diff
@@ -293,115 +296,168 @@ int mbedtls_des3_crypt_ecb( mbedtls_des3_context *ctx,
-    uint32_t val;
-    volatile uint32_t *tdes_key = (uint32_t *) ((uint32_t) &CRPT->TDES0_KEY1H + (0x40 * 0));
-    val = nu_get32_be(key[0] + 0);
-    *tdes_key ++ = val;
-    val = nu_get32_be(key[0] + 4);
-    *tdes_key ++ = val;
-    val = nu_get32_be(key[1] + 0);
-    *tdes_key ++ = val;
-    val = nu_get32_be(key[1] + 4);
-    *tdes_key ++ = val;
-    val = nu_get32_be(key[2] + 0);
-    *tdes_key ++ = val;
-    val = nu_get32_be(key[2] + 4);
-    *tdes_key ++ = val;
     
+    /* TODO: Change busy-wait to other means to release CPU */
+    /* Acquire ownership of DES H/W */
+    while (! crypto_des_acquire());
+    
+    /* Init crypto module */
+    crypto_init();
+    /* Enable DES interrupt */
+    TDES_ENABLE_INT();
+    
+    /* Configure TDES_CTL register
+     *
+     * BSP TDES driver supports multiple channels. Just use channel #0.
+     *
+     * Relationship of keying option and TDES H/W mode configuration
+     * 1: All three keys are independent                    ==> TDES 3-key mode (TMODE=1, 3KEYS=1)
+     * 2: K1 and K2 are independent, and K3 = K1            ==> TDES 2-key mode (TMODE=1, 3KEYS=0)
+     * 3: All three keys are identical, i.e. K1 = K2 = K3   ==> DES mode (TMODE=0)
+     *
+     * tdes_opmode is combination of TMODE/OPMODE, but TDES_Open I/F requires TMODE/OPMODE to be separate.
+     * We need to divide tdes_opmode to TMODE and OPMODE.
+     *
+     * TDES_IN_OUT_WHL_SWAP lets TDES H/W know input/output data are arranged in below for DMA transfer:
+     * 1. BE for byte sequence in word
+     * 2. BE for word sequence in double-word
+     */
+    TDES_Open(0,                                                // Channel number (0~4) 
+            enc,                                                // 0: decode, 1: encode
+            (tdes_opmode & CRPT_TDES_CTL_TMODE_Msk) ? 1 : 0,    // 0: DES, 1: TDES 
+            (keyopt == 1) ? 1 : 0,                              // 0: TDES 2-key mode, 1: TDES 3-key mode
+            tdes_opmode & CRPT_TDES_CTL_OPMODE_Msk,             // ECB/CBC/CFB/OFB/CTR
+            TDES_IN_OUT_WHL_SWAP);                              // TDES_NO_SWAP~TDES_IN_OUT_WHL_SWAP
+
+    /* Set DES/TDES keys 
+     *
+     * TDES_SetKey requires 3x2 word array. Change 3x8 byte array to 3x2 word array.
+     */
+    unsigned i;
+    uint32_t keys3x2[3][2];
+    for (i = 0; i < 3; i ++ ) {
+        keys3x2[i][0] = nu_get32_be(key[i] + 0);
+        keys3x2[i][1] = nu_get32_be(key[i] + 4);
+    }
+    TDES_SetKey(0, keys3x2);
+
     uint32_t rmn = length;
     const unsigned char *in_pos = input;
     unsigned char *out_pos = output;
-    
-    while (rmn) {
+
+    while (rmn > 0) {
         uint32_t data_len = (rmn <= MAXSIZE_DMABUF) ? rmn : MAXSIZE_DMABUF;
-        
+
         uint32_t ivh, ivl;
         ivh = nu_get32_be(iv);
         ivl = nu_get32_be(iv + 4);
         TDES_SetInitVect(0, ivh, ivl);
-    
+
         memcpy(dmabuf_in, in_pos, data_len);
```

Comment:
@ccli8 Is there a reason why you're always operating on the static DMA buffer for DES while attempting to use the user-provided buffer for AES (if it is in the DMA-region)?

Reply:
@hanno-arm They are coded by different ones. One for simplicity and the other for performance.